src/multiselect.js

```diff
@@ -0,0 +1,206 @@
+
+    this.controls_ = new MultiselectControls(
+        this.workspace_, options.multiselectIcon, this);
+    if (!options.multiselectIcon || !options.multiselectIcon.hideIcon) {
+      const svgControls = this.controls_.createDom();
+      this.workspace_.getParentSvg().appendChild(svgControls);
+    }
+    this.controls_.init();
+
+    if (options.useDoubleClick) {
+      this.useDoubleClick_(true);
+    }
+
+    if (options.baseBlockDragger) {
+      setBaseBlockDragger(options.baseBlockDragger);
+    }
+
+    if (!options.bumpNeighbours) {
+      this.origBumpNeighbours = Blockly.BlockSvg.prototype.bumpNeighbours;
+      Blockly.BlockSvg.prototype.bumpNeighbours = function() {};
+    }
+  }
+
+  /**
+   * Unbind the events and replace with original registration.
+   */
+  dispose() {
+    if (this.onKeyDownWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyDownWrapper_);
+      this.onKeyDownWrapper_ = null;
+    }
+    if (this.onKeyUpWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyUpWrapper_);
+      this.onKeyUpWrapper_ = null;
+    }
+    if (this.eventListenerWrapper_) {
+      this.workspace_.removeChangeListener(this.eventListenerWrapper_);
+      this.eventListenerWrapper_ = null;
+    }
+
+    ContextMenu.unregisterContextMenu();
+    ContextMenu.registerOrigContextMenu();
+    Blockly.ContextMenuRegistry.registry.unregister('workspaceSelectAll');
+    Shortcut.unregisterShortcut();
+    Blockly.ShortcutRegistry.registry.unregister('selectall');
+    Shortcut.registerOrigShortcut();
+
+    if (this.controls_) {
+      this.controls_.dispose();
+      this.controls_ = null;
+    }
+
+    this.useDoubleClick_(false);
+
+    if (this.origBumpNeighbours) {
+      Blockly.BlockSvg.prototype.bumpNeighbours = this.origBumpNeighbours;
+    }
+  }
+
+  /**
+   * Add double click to expand/collapse blocks (MIT App Inventor Supported).
+   * @param {!boolean} on Whether to turn on the mode.
+   * @private
+   */
+  useDoubleClick_(on) {
+    if (!on) {
+      Blockly.Gesture.prototype.handleWsStart = this.origHandleWsStart_;
+      return;
+    }
+
+    Blockly.Gesture.prototype.handleWsStart = (function(func) {
+      if (func.isWrapped) {
+        return func;
+      } else {
+        const wrappedFunc = function(e, ws) {
+          func.call(this, e, ws);
+          if (this.targetBlock_ && e.buttons === 1 &&
+              !inMultipleSelectionMode) {
+            const preCondition = function(block) {
+              return !block.isInFlyout && block.isMovable() &&
+              block.workspace.options.collapse;
+            };
+            if (Blockly.selected && preCondition(Blockly.selected)) {
+              if (ws.doubleClickPid_) {
+                clearTimeout(ws.doubleClickPid_);
+                ws.doubleClickPid_ = undefined;
+                if (Blockly.selected.id === ws.doubleClickBlock_) {
+                  const state = !Blockly.selected.isCollapsed();
+                  const apply = function(block) {
```

Comment:
`apply` could be confusing. Consider `maybeCollapse` or something similar so the function behavior is more clear from the name.